### ISSUE TITLE
Strange behaviour of NavigationAgent3D around NavigationObstacle3D

### ISSUE THREAD

**Issue by sadovsf** [godotengine/godot#67712]
### Godot version

4.0.beta3

### System information

macOS Monterey 12.6

### Issue description

When navigation agent should go around moving obstacles it almost seems it pushes into obstacle until it is no longer there (i tested it by disabling collisions on them but visually agent behaved the same). I would expect agent to try to "get around" dynamic obstacle rather than wait for it to clear.

### Steps to reproduce

Option 1:
Have agent go around 2 obstacles that are moving up and down (not sure direction has effect) each being down at different time. Their movement is done using animation player animating position attribute of parent node like so:
![Snímek obrazovky 2022-10-21 v 15 13 32](https://user-images.githubusercontent.com/2211533/197204211-726c0433-b9b3-497a-bf93-9f4233afc02a.png)

You will notice agent trying to get "through" obstacle rather than go around it.

Option 2:
Open repro project, start it and you can order red ball (agent) to go around level by lef mouse clicking on floor (clicking on purple walls and ramps will not react, only floor does).

### Minimal reproduction project

[Navmesh.zip](https://github.com/godotengine/godot/files/9839242/Navmesh.zip)

**Comment by smix8** [comment]
Avoidance != Pathfinding

Avoidance does not change the pathfinding only the velocity. Your path insists that the agent moves through the obstacles und enforces this every single _physics_frame update.

3D avoidance shapes are radius / spheres. Your box / rect shaped obstacles only have this shape for visuals or physics collision but for avoidance calculation they are in fact very large spheres that only leave tiny gaps for the agent to move through when they are "up".

When the agent is close to the obstacle and the obstacle sphere moves down it "squeezes" the agent. The correct 3D avoidance would be to move through the floor but it obviously cant so it gives up with the suggested avoidance "safe_velocity" and moves on. If you have "ignore_y" enabled the safe_velocity will ignore the y-axis, so the agent basically stands still and waits for better times or until the obstacles moves its radius out of the overlap with the agent.

**Comment by sadovsf** [comment]
I understand all this (i was author of c++ navigation integration for one of AAA titles released not long ago). What seems to be big problem for me is that avoidance sphere cannot be seen. Maybe drawing them at least as part of navmesh debug rendering, ideally even in editor would help a lot.

Btw so avoidance is using navmesh itself for avoiding obstacles (seems it is not the case) but then again, i dont see the sphere so it is har to debug what system see internally 🤔

Great work btw!

**Comment by Zireael07** [comment]
> Maybe drawing them at least as part of navmesh debug rendering, ideally even in editor would help a lot.

Good idea <3

**Comment by sadovsf** [comment]
Btw i tried to add spheres roughly size that would encapsulate whole obstacle. There is visible enough space for agent to fit through when obstacle is up but it tries to push itself into it instead of going around using navmesh. Hope it helps with debugging.

![image](https://user-images.githubusercontent.com/2211533/197509596-c80f29ee-a617-4ed3-ba09-1c85bd0f338d.png)

**Comment by smix8** [comment]
Yes, it is difficult to find good avoidance property values without any debug visuals. Debug visuals for the editor are on the "to-do" list. Even with good property values and setup it is difficult with the current avoidance to get a good result.

The RVO2-3D library currently integrated in Godot is intended for air, water or space agents and has issues with flat surface 2D avoidance in general as soon as things get more crowded or something pushes from above or below. It also has no support for hard-surface obstacles to build e.g. navmesh constrains. All current NavigationObstacles are just a (soft) avoidance radius created by a fake agent with some properties set at values that it can push most other agents but not all.

These are issues that are outside of what a user can control or change without digging into the source code. As shown in the linked rvo rework proposal there are a few things currently worked on that should improve the avoidance situation for Godot 4.

**Comment by sadovsf** [comment]
I see, thank you for great details!😊 

Will be looking forward for further improvements then 👍

**Comment by smix8** [comment]
The new avoidance is more detailed without much jitter if in close contact with an obstacle even when following the contour of moving avoidance obstacles very closely which should also help this issue here. There is no collision in this animation here, this is only avoidance movement.

![rvo_obstacles8](https://user-images.githubusercontent.com/52464204/197746916-e70d6d42-d4f5-4327-b51f-bd990a75de86.gif)

**Comment by sadovsf** [comment]
Looks nice! Will test it as soon as it is available in beta build 👍
